src/monodroid/jni/monodroid-glue.c

```diff
@@ -535,6 +536,28 @@ get_xamarin_android_msbuild_path (void)
 	return msbuild_folder_path;
 }
 
+static char *libmonoandroid_directory_path = NULL;
+
+// Returns the directory in which this library was loaded from
+static char*
+get_libmonoandroid_directory_path ()
+{
+	wchar_t module_path[MAX_PATH];
+	char *converted_path = NULL;
+	HMODULE module = NULL;
+
+	if (libmonoandroid_directory_path != NULL)
+		return libmonoandroid_directory_path;
+
+	if (!GetModuleHandleEx (GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS | GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT,
+	                        (LPCSTR)&libmonoandroid_directory_path, &module))
+		return NULL;
+	GetModuleFileNameW (module, module_path, sizeof (module_path));
+	converted_path = utf16_to_utf8 (module_path);
+	libmonoandroid_directory_path = dirname (converted_path);
```

Comment:
Does Windows *provide* a `dirname()` function? I can't find documentation for it.

[I see mingw source code for `dirname()`](https://sourceforge.net/p/mingw-w64/mingw-w64/ci/907975aa516373f949fb773de43c6ba3642fb731/tree/mingw-w64-crt/misc/dirname.c), but I don't think we link against any mingw libraries -- do we? *Should* we?

I suspect the easier path would be to use [`strrchr()`](https://msdn.microsoft.com/en-us/library/ftw0heb9.aspx), akin to:

    *strrchr (converted_path, '\\') = '\0';
    libmonoandroid_directory_path = converted_path;

The primary reason I ask is because I don't know if `dirname()` allocates a new string or not, and I'm thinking about memory leaks. 

The [macOS **dirname**(3) man page](https://developer.apple.com/legacy/library/documentation/Darwin/Reference/ManPages/man3/dirname.3.html) is horrifying:

> The `dirname()` function returns a pointer to internal storage space allocated on the first call that will be overwritten by subsequent calls.  `dirname_r()` is therefore preferred for threaded applications.
> 
> Other vendor implementations of `dirname()` may modify the contents of the string passed to `dirname()`; if portability is desired, this should be taken into account when writing code which calls this function.

*Both* of those notes scream "flee in terror" to me...